### ISSUE TITLE
normalize_estimator function trying to access non-existent attribute

### ISSUE THREAD

**Issue by blinickg** [dmlc/xgboost#5687]
See thread here: dask/dask-ml#667

Proposing a change to subclass `AttributeError` to make it clear that you're accessing a non-existent attribute when calling the `normalize_estimator` function before training.

**Comment by trivialfis** [comment]
Em.. is there any specification for that?

**Comment by blinickg** [comment]
@trivialfis , is there enough specification at https://github.com/dask/dask-ml/issues/667  ?

**Comment by trivialfis** [comment]
Surely I can do that.  I'm just thinking about what's the easiest and future proof way to be fully compatible with scikit learn as situations like this comes up a lot.

**Comment by blinickg** [comment]
OK, let me know if I can provide further information.

**Comment by TomAugspurger** [comment]
> I'm just thinking about what's the easiest and future proof way to be fully compatible with scikit learn as situations like this comes up a lot.

You might try running `sklearn.utils.estimator_checks.check_estimator` on the XGBoost estimators: https://scikit-learn.org/stable/modules/generated/sklearn.utils.estimator_checks.check_estimator.html

**Comment by trivialfis** [comment]
Thanks!  Work on this is already on the way.  There are some tests in there we might not be able to pass without making a breaking change.  But I succeeded at passing most of the other tests.

**Comment by blinickg** [comment]
@trivialfis, did you manage to pass the tests? I'm just asking because my team is using the function so we're waiting for it to be fixed before we continue.

**Comment by trivialfis** [comment]
Yes, except for kwargs in __init__.  Will submit a PR this week.

**Comment by trivialfis** [comment]
A few of them are already mainlined actually, including the use of 'NotFittedError'.

**Comment by trivialfis** [comment]
I think the exception type is already fixed in master branch with `NotFittedError` replacing `XGBoostError`.  Let's open new issues for different errors.  I will also reply on https://github.com/dask/dask-ml/issues/667 when needed.
